### PR TITLE
Evaporative cooler RH control

### PR DESCRIPTION
**Pull request by yujiex**
Pull request overview
---------------------
# Introduction
The introduction of excessive moisture is one of the potential issues of direct evaporative coolers. A humidity control could become useful in preventing the evaporative cooler from raising indoor humidity to an uncomfortable level. Currently, the direct evaporative cooler can be controlled with the sensor node temperature using AvailabilityManagers (*AvailabilityManager:LowTemperatureTurnOff* or *AvailabilityManager:HighTemperatureTurnOn*). This feature proposes to add a relative humidity (RH) control to shut down the evaporative cooler when the indoor RH is too high. This is done by adding a `Shut Off Relative Humidity` in `ZoneHVAC:EvaporativeCoolerUnit` object.

# Testing with idf testfile

A test idf *StripMallZoneEvapCoolerRHcutoff.idf* is constructed by adapting from *StripMallZoneEvapCooler.idf*, with an added `Shut Off Relative Humidity` in `LGstore1 Zone Evap Unit`.
First, we confirmed that with the added field, when the `Shut Off Relative Humidity = 100`, the evaporative cooler energy consumption is the same as the *StripMallZoneEvapCooler.idf* when the field is not present.
cmp_no_cutoff_vs_cutoff_100

Second, when the `Shut Off Relative Humidity = 0`, the evaporative cooler energy consumption is zero.
cmp_100_cutoff_vs_cutoff_0

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: design/FY2023/NFP-evapCoolerRHcontrol.md
Comment (rraustad) [inline design/FY2023/NFP-evapCoolerRHcontrol.md]: Since one of the temperatures used to control this type of equipment would be zone temperature, which changes seasonally, this temperature field should probably be a schedule. Using that same reasoning, all these air condition inputs could be schedules for added flexibility.

Comment (yujiex) [reply design/FY2023/NFP-evapCoolerRHcontrol.md]: make sense. We can discuss this in the technicality meeting.

**Comment by mjwitte** [comment]
@yujiex Looking at this again
1. I'm ok with adding a new availability manager, for now. 
2. Long-term, a ZoneHVAC evap cooler could be useful to simplify this. 
3. Instead of adding to the growing list of availability managers, would you be open to adding one general availability manager? e.g.
```
AvailaibilityManager:SystemNodeConditions,
 A1 , \field Name
      \required-field
      \type alpha
      \reference SystemAvailabilityManagers
 A2,  \field Action Type
      \required-field
      \type choice
      \key TurnOn
      \key TurnOff
 A3 , \field Sensor Node Name
      \note The system node where the conditions are monitored.
      \required-field
      \type node
 A4 , \field Limit 1 Condition
      \required-field
      \type choice
      \key Temperature
      \key HumidityRatio
      \key DewPoint
      \key RelativeHumidity
 A5,  \field Limit 1 Type
      \required-field
      \type choice
      \key Maximum
      \key Minimum
 A6,  \field Limit 1 Schedule Name
      \required-field
      \type object-list
      \object-list ScheduleNames
 A7 , \field Limit 2 Condition
      \type choice
      \key None
      \key Temperature
      \key HumidityRatio
      \key DewPoint
      \key RelativeHumidity
      \default None
 A8, \field Limit 2 Type
     \type choice
     \key None
     \key Maximum
     \key Minimum
     \default None
 A9, \field Limit 2 Schedule Name
     \type object-list
     \object-list ScheduleNames
```

**Comment by yujiex** [comment]
> @yujiex Looking at this again
> 
> 1. I'm ok with adding a new availability manager, for now.
> 2. Long-term, a ZoneHVAC evap cooler could be useful to simplify this.
> 3. Instead of adding to the growing list of availability managers, would you be open to adding one general availability manager? e.g.
> 
> ```
> AvailaibilityManager:SystemNodeConditions,
>  A1 , \field Name
>       \required-field
>       \type alpha
>       \reference SystemAvailabilityManagers
>  A2,  \field Action Type
>       \required-field
>       \type choice
>       \key TurnOn
>       \key TurnOff
>  A3 , \field Sensor Node Name
>       \note The system node where the conditions are monitored.
>       \required-field
>       \type node
>  A4 , \field Limit 1 Condition
>       \required-field
>       \type choice
>       \key Temperature
>       \key HumidityRatio
>       \key DewPoint
>       \key RelativeHumidity
>  A5,  \field Limit 1 Type
>       \required-field
>       \type choice
>       \key Maximum
>       \key Minimum
>  A6,  \field Limit 1 Schedule Name
>       \required-field
>       \type object-list
>       \object-list ScheduleNames
>  A7 , \field Limit 2 Condition
>       \type choice
>       \key None
>       \key Temperature
>       \key HumidityRatio
>       \key DewPoint
>       \key RelativeHumidity
>       \default None
>  A8, \field Limit 2 Type
>      \type choice
>      \key None
>      \key Maximum
>      \key Minimum
>      \default None
>  A9, \field Limit 2 Schedule Name
>      \type object-list
>      \object-list ScheduleNames
> ```

@mjwitte  Thanks for the comments during the meeting. I agree that the ZoneHVAC equipment is a nicer option in the long term. A generic availability manager object is nice too. Should I do it for the two relative humidity-related ones for now? Or do the existing ones also need to convert to (or be compatible with) using this generic one as well?

**Comment by mjwitte** [comment]
@yujiex My thought was by adding a general manager like `AvailaibilityManager:SystemNodeConditions` then you would not need to add specific RH availability managers. All existing managers can stay in place. I would like some other opinions on this @Myoldmopar @rraustad

**Comment by rraustad** [comment]
A generic availability manager is no different than pulling packaged terminal units into UnitarySystem. One place for code changes. I think this is a good idea for availability managers. There are just too many flavors now to add even more. Whether the availability manager should take the role of RH control for evap coolers is another question. The evap cooler is already set up to do some controls on temperature. I guess one could argue that temperature is the primary control point for a basic evap cooler. Other custom controls could be managed through availability managers as @yujiex has suggested. The benefit of using avail managers is that they could be useful for other equipment types. I think using an avail manager is the best path forward and use of a generic manager may lead to a reduction in the number of existing avail managers.

    ZoneHVAC:EvaporativeCoolerUnit,
        \memo Zone evaporative cooler. Forced-convection cooling-only unit with supply fan,
        \memo 100% outdoor air supply. Optional relief exhaust node
        \min-fields 15
      A1 , \field Name
        \required-field
        \reference ZoneEquipmentNames
      A2 , \field Availability Schedule Name
        \note Availability schedule name for this system. Schedule value > 0 means the system is available.
        \note If this field is blank, the system is always available.
        \type object-list
        \object-list ScheduleNames
      A3,  \field Availability Manager List Name
        \note Enter the name of an AvailabilityManagerAssignmentList object.
        \type object-list
        \object-list SystemAvailabilityManagerLists

      A10, \field Cooler Unit Control Method
        \required-field
        \type choice
        \key ZoneTemperatureDeadbandOnOffCycling
        \key ZoneCoolingLoadOnOffCycling
        \key ZoneCoolingLoadVariableSpeedFan
      N2 , \field Throttling Range Temperature Difference
        \note used for ZoneTemperatureDeadbandOnOffCycling hystersis range for thermostatic control
        \type real
        \units deltaC
        \default 1.0
        \minimum> 0.0
      N3 , \field Cooling Load Control Threshold Heat Transfer Rate
        \type real
        \units W
        \default 100.0
        \note Sign convention is that positive values indicate a cooling load
        \minimum> 0.0


    AvailabilityManagerAssignmentList,
       \key AvailabilityManager:Scheduled
       \key AvailabilityManager:ScheduledOn
       \key AvailabilityManager:ScheduledOff
       \key AvailabilityManager:NightCycle
       \key AvailabilityManager:DifferentialThermostat
       \key AvailabilityManager:HighTemperatureTurnOff
       \key AvailabilityManager:HighTemperatureTurnOn
       \key AvailabilityManager:LowTemperatureTurnOff
       \key AvailabilityManager:LowTemperatureTurnOn
       \key AvailabilityManager:NightVentilation
       \key AvailabilityManager:OptimumStart

**Comment by mjwitte** [comment]
I didn't realize we already have `ZoneHVAC:EvaporativeCoolerUnit`. So why aren't we just adding more controls to that instead of adding availability managers?

**Comment by rraustad** [comment]
> So why aren't we just adding more controls to that instead of adding availability managers?

Because an availability manager could handle that role and also other roles for other equipment. Why should there be code in evap coolers to handle RH controls, and code in furnace to handle RH control, and ... I originally thought it would be better to add to evap coolers, but the more I think about it having code in 1 place and used whereever needed is more efficient. @Myoldmopar?

**Comment by yujiex** [comment]
Thanks for the comments @mjwitte @rraustad. We'll do the following modification

1. create a generic availability manager, which can be used in specifying the relative humidity-related availability. For the object design
- maybe we can add an "enthalpy" to the "Limit 1 Condition" and "Limit 2 Condition"?
- is "DewPoint" or "DewPointTemperature" better?
2. I also didn't realize there's a zone equipment evaporative cooler already. Somehow missed it in the search. Thanks @rraustad for pointing it out.
3. re @mjwitte 's comment about why not adding more controls to the `ZoneHVAC:EvaporativeCoolerUnit`. The control methods in the objects are the following. 

```
   A10, \field Cooler Unit Control Method
        \required-field
        \type choice
        \key ZoneTemperatureDeadbandOnOffCycling
        \key ZoneCoolingLoadOnOffCycling
        \key ZoneCoolingLoadVariableSpeedFan
```

The humidity control interacts with all of them, so we cannot directly add a fourth control method here. For example, when the "ZoneCoolingLoadOnOffCycling" method is selected here, maybe the cooling load is high enough for the cooler to operate but the indoor condition is too humid, then the system should still be off. This means either we'll have the relative humidity control built into all of the three existing control methods, or use the availability manager to shut it off. I felt the availability manager way is cleaner. The object `ZoneHVAC:EvaporativeCoolerUnit` has a `Availability Manager List Name` field, so we can just reference the newly added generic availability manager object to specify the relative humidity-based system availability. So I'm agreeing with @rraustad on using availability manager for the relative humidity control.

**Comment by rraustad** [comment]
@yujiex also remember that it is an Availability Manager List, with more than 1 possible managers listed. As this description shows, there can be a priority order or a first to act methodology.

    1.26.9 AvailabilityManagerAssignmentList

    The AvailabilityManagerAssignmentList defines the applicable managers used for an AirLoopHVAC or
    PlantLoop. The priority of availability managers is based on a set of rules and are specific to the type of
    loop. The output from each Availability Manager is an availability status flag. This flag can have the values
    NoAction, ForceOff, CycleOn, or CycleOnZoneFansOnly (used only for air loops). The availability status
    flags for the Availability Managers referenced by an air or plant loop are used to set the availability status
    flag for each loop. For the air loops, ForceOff takes precedence: if any of the loop’s availability managers
    are showing status ForceOff, the loop status will be ForceOff. Next in precedence is CycleOnZoneFansOnly,
    followed by CycleOn, and NoAction. For the plant loops, there is no precedence among the Availability
    Manager status flag values. Instead, the first availability manager giving a status flag value other than
    NoAction sets the status for the loop. The Availability Managers are executed in Availability Manager List
    order.

**Comment by yujiex** [comment]
@rraustad I guess for the current application, since "Availability Managers are executed in Availability Manager List
order", the system-always-off case can be controlled with two availability managers: high-relative-humidity-turn-off, followed by high-temperature-turn-on? In this case, a generic availability manager with one set of control condition (A4-A6 in the idd object Mike suggested) might be enough. But maybe it's still nice to have a generic one with two sets of conditions, as the sequential one does not cover the second row in the following truth table? @rraustad @mjwitte thoughts?

|P|Q|P and Q|
| :---: | :---: | :---: |
|T|T|T|
|T|F|F|
|F|T|F|
|F|F|F|

**Comment by rraustad** [comment]
Your table works if P=T and Q=RH or P=RH and Q=T. So this is not priority, or first to act, it's cumulative. If any are false, then system is off.  What does it mean if there are multiple outcomes based on priority, order, or cumulative? What does the user want? Sounds like an input choice to me.

**Comment by yujiex** [comment]
> Your table works if P=T and Q=RH or P=RH and Q=T. So this is not priority, or first to act, it's cumulative. If any are false, then system is off. What does it mean if there are multiple outcomes based on priority, order, or cumulative? What does the user want? Sounds like an input choice to me.

@rraustad P and Q are two statements that can be either true or false. For example, P could be something like `Temperature > 15`. 
For the use case of a relative humidity shutoff control, the on/off of the system is dictated by the relative humidity statement. Whenever it's above the threshold, the system will be off regardless of the temperature or cooling load condition.

But this type of availability manager can also implement something like this: `if (10 < Temperature) and (Temperature < 25) then turn the system on`. In this case, neither `(10 < Temperature)` nor `(Temperature < 25)` can dictate the final result of `((10 < Temperature) and (Temperature < 25))`. If we want to have this type of availability manager, we have to have two sets of conditions like A4-A9 in Mike's idd object.

**Comment by rraustad** [comment]
@yujiex for your case of a min and max you would have A5 = min and A8 = max. Don't overcomplicate this. Or since fields past A6 are not "required" you could have 2 avail managers, one min and one max. These would be cumulative. The more I think about this generic avail manager the more I like it. 

I don't mean to complicate this but if you add schedule as a key choice you could eliminate 7 legacy avail managers (4 temp and 3 schedule).

    A4 , \field Limit 1 Condition
         \required-field
         \type choice
         \key Temperature
         \key HumidityRatio
         \key DewPoint
         \key RelativeHumidity

         \key Schedule

**Comment by yujiex** [comment]
@rraustad For schedule, fields `A6,  \field Limit 1 Schedule Name` and `A9, \field Limit 2 Schedule Name` could take care of that. Anyway, I think we're all agreeing on Mike's formulation of the generic availability manager design.  

I saw the RelativeHumidity key now, my bad.

**Comment by rraustad** [comment]
@yujiex the schedule key would turn on or off a component based on the schedule value itself, not a comparison of that value to something else. You need the key choice to know what to do with the schedule.

**Comment by yujiex** [comment]
> @yujiex the schedule key would turn on or off a component based on the schedule value itself, not a comparison of that value to something else. You need the key choice to know what to do with the schedule.

Oh, I understand now. This schedule in the `Limit 1 Condition` is an On/Off directly. The schedule in A6 and A9 are threshold values.

**Comment by rraustad** [comment]
The ultimate generic avail manager would look like this but would require more input fields. This is not a suggestion to add these new key choices, this is a thought provoker only to add to this discussion.

     A4 , \field Limit 1 Condition
      \required-field
      \type choice
      \key Temperature
      \key HumidityRatio
      \key DewPoint
      \key RelativeHumidity
      \key Pressure
      \key Enthalpy
      \key Flow
      \key Schedule
      \key NightCycle
      \key DifferentialThermostat
      \key NightVentilation
      \key OptimumStart

**Comment by yujiex** [comment]
> The ultimate generic avail manager would look like this but would require more input fields. This is not a suggestion to add these new key choices, this is a thought provoker only to add to this discussion.
> 
> ```
>  A4 , \field Limit 1 Condition
>   \required-field
>   \type choice
>   \key Temperature
>   \key HumidityRatio
>   \key DewPoint
>   \key RelativeHumidity
>   \key Pressure
>   \key Enthalpy
>   \key Flow
>   \key Schedule
>   \key NightCycle
>   \key DifferentialThermostat
>   \key NightVentilation
>   \key OptimumStart
> ```

So many possibilities indeed!

**Comment by mjwitte** [comment]
If it takes this much thought to apply it correctly, then I regret suggesting the generic availability manager.  Use a Python plugin instead so the logic can be explicit.

If the goal here is to model in-zone evap coolers with a high-RH limit, then a couple of new fields in `ZoneHVAC:EvaporativeCoolerUnit` may be the best approach.  Something like
```
ZoneHVAC:EvaporativeCoolerUnit,
...
   N?, \field Shut Off Relative Humidity
        \note Zone relative humidity above which the evap cooler is shut off.
        \type real
        \minimum 0.00
        \maximum 100.00
        \units percent
   N?, \field Turn On Relative Humidity
        \note Zone relative humidity below which the evap cooler is allowed to turn on.
        \type real
        \minimum 0.00
        \maximum 100.00
        \units percent
```

**Comment by yujiex** [comment]
@mjwitte Yeah, for the zone-level evaporative cooler unit, `Shut Off Relative Humidity` and `Turn On Relative Humidity` is cleaner. Do you think we still want a generic `AvailaibilityManager:SystemNodeConditions` object in case people want to use the non-zone evaporative cooler?

Also a side-question here, for the ZoneHVAC:EvaporativeCoolerUnit, is it possible to configure it as recirculating and make its `Outdoor Air Inlet Node Name` field accept a zone node? Some of the portable ones we've see are not connected to outdoor.

**Comment by nrel-bot-2c** [comment]
@yujiex it has been 7 days since this pull request was last updated.

**Comment by mjwitte** [comment]
>Do you think we still want a generic AvailaibilityManager:SystemNodeConditions object in case people want to use the non-zone evaporative cooler?

@yujiex I would say no, but that's up to you and @hongtz68 to decide what's needed.

> Also a side-question here, for the ZoneHVAC:EvaporativeCoolerUnit, is it possible to configure it as recirculating and make its Outdoor Air Inlet Node Name field accept a zone node? Some of the portable ones we've see are not connected to outdoor.

Yes, but the field name should be changed to "Cooler Inlet Node Name" and this call [here](https://github.com/NREL/EnergyPlus/blob/225390b17343395c44d86a495f2979df416d4bb0/src/EnergyPlus/EvaporativeCoolers.cc#L3447-L3455) should change `DataLoopNode::ConnectionType::OutsideAir` to `DataLoopNode::ConnectionType::Inlet`. And (at some point) the struct variable `OAInletNodeNum` should change to `UnitInletNodeNum` (but that will introduce many nuisance code changes, so that could wait until this PR is almost complete).

**Comment by yujiex** [comment]
> > Do you think we still want a generic AvailaibilityManager:SystemNodeConditions object in case people want to use the non-zone evaporative cooler?
> 
> @yujiex I would say no, but that's up to you and @hongtz68 to decide what's needed.
> 

Thanks @mjwitte. I'll discuss this with Tianzhen today.
> > Also a side-question here, for the ZoneHVAC:EvaporativeCoolerUnit, is it possible to configure it as recirculating and make its Outdoor Air Inlet Node Name field accept a zone node? Some of the portable ones we've see are not connected to outdoor.
> 
> Yes, but the field name should be changed to "Cooler Inlet Node Name" and this call [here](https://github.com/NREL/EnergyPlus/blob/225390b17343395c44d86a495f2979df416d4bb0/src/EnergyPlus/EvaporativeCoolers.cc#L3447-L3455) should change `DataLoopNode::ConnectionType::OutsideAir` to `DataLoopNode::ConnectionType::Inlet`. And (at some point) the struct variable `OAInletNodeNum` should change to `UnitInletNodeNum` (but that will introduce many nuisance code changes, so that could wait until this PR is almost complete).

Thanks for the clarification. I agree we can leave this to when the PR is mostly done.

**Comment by yujiex** [comment]
I was wondering if we need a low-relative-humidity-turning-on option as well or just one high-humidity cutoff. When the humidity is low, should the on/off be controlled by cooling needs or temperature? Otherwise, maybe it could over-cool the zone. What do you feel @mjwitte?

File: design/FY2023/NFP-evapCoolerRHcontrol.md
Comment (rraustad) [inline design/FY2023/NFP-evapCoolerRHcontrol.md]: These notes should be revised. When "Maximum" is chosen, then the action is taken when the value of A4 is greater than the threshold. When "Minimum" ... less than ... Also, doesn't it depend on which type of equipment uses this avail manager and then some combinations would not make sense. What if a dehumidifier using RelativeHumidity tried to use TurnOff and Maximum? or TurnOn and Minimum? Or if the EvapCooler got the inputs backwards, as RelativeHumidity, TurnOff and Minimum. Would these backward inputs be ignored or just do what the user specifies? I guess doing what the user specified would make the most sense since the avail manager does not know what type of equipment is being controlled.

Comment (yujiex) [reply design/FY2023/NFP-evapCoolerRHcontrol.md]: @rraustad per @mjwitte 's suggestion, we decided to leave out the generic availability manager and only modify the ZoneHVAC object in this PR.

**Comment by mjwitte** [comment]
> I was wondering if we need a low-relative-humidity-turning-on option as well or just one high-humidity cutoff. When the humidity is low, should the on/off be controlled by cooling needs or temperature? Otherwise, maybe it could over-cool the zone. What do you feel @mjwitte?

I guess that depends if someone would use this as a humidifier. That seems unlikely to me, but I don't have much familiarity with how this equipment is used.

**Comment by yujiex** [comment]
> > I was wondering if we need a low-relative-humidity-turning-on option as well or just one high-humidity cutoff. When the humidity is low, should the on/off be controlled by cooling needs or temperature? Otherwise, maybe it could over-cool the zone. What do you feel @mjwitte?
> 
> I guess that depends if someone would use this as a humidifier. That seems unlikely to me, but I don't have much familiarity with how this equipment is used.

When it is used as a humidifier, should we add an option to the `Cooler Unit Control Method` field to allow for humidity to take priority over the temperature/load-based control?

**Comment by hongtz68** [comment]
> 

Let us keep this object clean focusing on evaporative cooling. Humidification is done in another object/way.

**Comment by yujiex** [comment]
> > 
> 
> Let us keep this object clean focusing on evaporative cooling. Humidification is done in another object/way.

sure. I think then we'll just need the high-temperature cutoff.

**Comment by nrel-bot-2b** [comment]
@yujiex it has been 7 days since this pull request was last updated.

**Comment by yujiex** [comment]
Hi @mjwitte and @rraustad, the PR is ready for review. Would you mind taking another look? Thanks :)

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (Myoldmopar) [inline doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:1554]: Straightforward :+1:

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in:37021]: This is a new optional field, so I assume the code below will check for blank.  I see in the IDD we sometimes have "Relative Humidity Fraction" from 0 to 1, but we also have it as a percent from 0 to 100.  So this is fine.

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EvaporativeCoolers.cc:2934]: Is this just a comment-only change?  I don't see anything different besides it split into two lines.

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EvaporativeCoolers.cc:3609]: OK, so if it is blank, it stays as the default 100.  This seems reasonable ... though I guess there is a possibility that some user file is currently still dumping moisture into the space even when the RH is at 100, and now with this change we are quietly turning it off.  The other option is to keep a bool to decide if this field was blank, and if so, use the exact same behavior as before.  I'm not suggesting we do that, just walking myself through the changes.  Leave it, this is fine.

File: src/EnergyPlus/EvaporativeCoolers.hh
Comment (Myoldmopar) [inline src/EnergyPlus/EvaporativeCoolers.hh:310]: Although the IDD object inputs the RH as a percent, I would have preferred we keep it as a fraction internally, to make the calculations one step easier.  I'm not going to request that change now because it looks like this could be done or almost done.  But in general, I think percentages should just be a user-facing thing and keep things in fractional form internally where reasonable.

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (Myoldmopar) [inline src/EnergyPlus/EvaporativeCoolers.cc:4062]: I assume this is fine.  As long as all the unit outputs and node connections are cleaned up, then we are able to return early.  I only bring it up because returning early is a common source of bugs in component models.  In this case, since you are calling Calc, I think it should be OK.

File: testfiles/CMakeLists.txt
Comment (Myoldmopar) [inline testfiles/CMakeLists.txt:576]: New test file :+1:

File: tst/EnergyPlus/unit/ZoneHVACEvaporativeCooler.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ZoneHVACEvaporativeCooler.unit.cc:644]: Unit test looks good.

**Review by Myoldmopar — APPROVED** [review]
I had a few comments, but nothing that should hold this up.  This looks good, and I'm marking it approved.

**Comment by Myoldmopar** [comment]
@mjwitte or @rraustad any further comments/concerns here?  This looks ready to go.

**Comment by rraustad** [comment]
The new model appears to be working with limited code changes. I am OK with merging.

**Review by mjwitte — APPROVED** [review]
I approve, based on a quick idd and code review.

File: doc/input-output-reference/src/overview/group-zone-forced-air-units.tex
Comment (yujiex) [reply doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:1554]: :)

File: src/EnergyPlus/EvaporativeCoolers.hh
Comment (yujiex) [reply src/EnergyPlus/EvaporativeCoolers.hh:310]: make sense.

File: tst/EnergyPlus/unit/ZoneHVACEvaporativeCooler.unit.cc
Comment (yujiex) [reply tst/EnergyPlus/unit/ZoneHVACEvaporativeCooler.unit.cc:644]: :)

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (yujiex) [reply src/EnergyPlus/EvaporativeCoolers.cc:3609]: okay cool

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (yujiex) [reply src/EnergyPlus/EvaporativeCoolers.cc:4062]: is it better to keep a flag and return at the end? I'll take a note here and do that in the future. @Myoldmopar

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (yujiex) [reply src/EnergyPlus/EvaporativeCoolers.cc:2934]: I think it just reformatted the code to two lines. Sometimes the clang-format tries to align the starting point of the comments. So even if this particular line is not changed, sometimes the formatting wants it differently based on changes of other lines. That might be what's happening here.

File: src/EnergyPlus/EvaporativeCoolers.cc
Comment (Myoldmopar) [reply src/EnergyPlus/EvaporativeCoolers.cc:4062]: It just depends.  In the past, returning early was a common way for nodes to not get updated, leading to many bugs in E+.  Those days are mostly behind us, but returning early is still something that jogs the memory.  As long as the component "finishes" its work, then go ahead and return, no problem.

**Pull request merged**
